Add long issue list option

vyrondrosos opened the pull request.

dspinellis commented:
Thank you!  Please fix the failing test, and squash the commits into a single one.  I will then review the code.

vyrondrosos commented:
I've updated my repository.

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+  Adate=`git show --no-patch --format='%ai' $id`
+
+  # Milestone
+  if [ -s $path/milestone ] ; then
```
Please add double quotes around path

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+
+  # Milestone
+  if [ -s $path/milestone ] ; then
+    Amilestone=`fmt $path/milestone`
```
Use the modern $( rather than `

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+  Adescription=`head -n 1 $path/description` 
+
+
+  (echo -nE "$sortfield";printf '\002';echo -E "$formatstring") | #we print the field to sort by first, and remove it after sorting 
```
Simply echo "$sortfield"$'\002'"$formatstring"

dspinellis replied:
Write # Print the field
(Space after # and start with a capital letter.)

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+  -e s$'\003'%M$'\003'"$Amilestone"$'\003'g \
+  -e s$'\003'%A$'\003'"$Aassignee"$'\003'g \
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
```
Place the comment before the sed invocation; Write: # An ...

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+      ;;
+    o)
+      sortfield=$OPTARG
+      if [ -z $sortfield ] ; then
```
Add double quotes

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
       ;;
+    l)
+      long=1
+      formatstring=$OPTARG
```
Add double quotes

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+      if [ -z $sortfield ] ; then
+        usage_list
+      fi
+      if echo $sortfield | grep -v -E '^%[icMATD]$' > /dev/null; then
```
I think you can write this with expr(1)

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
 
+  case $formatstring in
+    oneline)
+      formatstring='ID:%i  DATE:%c  TAGS:%T  DESC:%D'
```
Shouldn't the words be capital case, e.g. Date?
Also space after :

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
+  tr "\n" '\001' >> results
+  printf "\n" >> results
```
echo is enough

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,97 @@ usage_list()
+  -e s$'\003'%A$'\003'"$Aassignee"$'\003'g \
+  -e s$'\003'%T$'\003'"$Atags"$'\003'g \
+  -e s$'\003'%D$'\003'"$Adescription"$'\003'g | #an nonprintable char is used as delimiter since the fields can contain any char
+  tr "\n" '\001' >> results
```
Don't append do the file. Instead write to stdout, and have a single tee capture the output. (See below.)

dspinellis commented on git-issue.sh:
```diff
@@ -889,13 +960,22 @@ sub_list()
+    while read path id ; do
+      shortshow
+    done 
+    cp results ~/res
```
Forgotten debug statement?

dspinellis commented on git-issue.sh:
```diff
@@ -889,13 +960,22 @@ sub_list()
+   
+    while read path id ; do
+      shortshow
+    done 
```
done |
sort $sortrev | 
sed 's/^.*\x02//' | 
tr '\001' "\n"

dspinellis commented on git-issue.sh:
```diff
@@ -889,13 +960,22 @@ sub_list()
+      head -1 $path/description
+    done |
+    sort -k 2 |
+    tee results |
```
Move tee results | pager outside the if statement.

dspinellis commented on git-issue.sh:
```diff
@@ -985,6 +1065,7 @@ Work with an issue
 
 Show multiple issues
    list       List open issues (or all with -a)
+* list -l formatstring: This will list issues in a specifiable format, given as an argument to -l
```
in the specified format.

vyrondrosos commented:
Thanks for the feedback. I made the changes, wrote a few tests, and squashed the commits again. However, it seems like the tests I added fail on the Travis CI system, while they pass on mine. Is that normal?

dspinellis commented:
It's not normal. Only one of the two test targets fails.  Could it be that you're using a non-standard sed feature?

```
sed: -e expression #1, char 17: unknown option to `s'
fail 85 - grep Second issue
sed: -e expression #1, char 17: unknown option to `s'
ok 86 - not grep First-issue
sed: -e expression #1, char 17: unknown option to `s'
sed: -e expression #1, char 17: unknown option to `s'
fail 87 - grep feature
```

vyrondrosos commented:
After doing some testing, it seems that the problem lies with the nonprintable characters I use as delimiters to `sed`(since the fields themselves, especially the issue's description, could contain any printable character). Not sure what the alternative would be though, apart from switching `sed` out for another tool.

vyrondrosos commented:
The implemented tests are now passing. If you have any ideas on further improvements, please let me know.

dspinellis commented:
Great!  The code review seems to be pending.

vyrondrosos commented:
Indeed, it seems I had missed some of your suggestions(the ones Github didn't show automatically). It should be fixed now.

dspinellis commented on README.md:
```diff
@@ -93,6 +93,27 @@ You use _git issue_ with the following sub-commands.
+   The following escape sequences can be used:
+   
+   - `%n` : newline
+   
```
Can we please loose the empty lines between the bullets?

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,98 @@ usage_list()
+
+  # Assignee
+  if [ -r "$path/assignee" ] ; then
+    assignee=$(fmt $path/assignee|sed -e 's/[\/&]/\\&/g')
```
Some more quotes in $path are needed.

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,98 @@ usage_list()
+  description=$(head -n 1 $path/description|sed -e 's/[\/&]/\\&/g')
+
+
+  (echo "$sortfield"$'\002'"$formatstring") | # Print the field to sort by first, and remove it after sorting 
```
Move the comment on the line above

dspinellis commented on git-issue.sh:
```diff
@@ -855,26 +855,98 @@ usage_list()
+
+  (echo "$sortfield"$'\002'"$formatstring") | # Print the field to sort by first, and remove it after sorting 
+
+  # Nonprintable character is used as delimiter since the fields can contain any printable character
```
This comment is no longer valid

dspinellis commented:
Thank you; almost there!  Just a few small changes remaining.

vyrondrosos commented:
Done.

dspinellis commented:
Well done, thank you!
